### PR TITLE
[limitranger] default resource policy validate request <= default.limit

### LINKED ISSUE #51430: default limit sometimes acts as max limit
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
I created a `LimitRange` with the following spec:
```
apiVersion: v1
kind: LimitRange
metadata:
  name: mem-limit-range
spec:
  limits:
  - default:
      memory: 512Mi
    defaultRequest:
      memory: 256Mi
    type: Container
```

Now,
- When I create a pod without specifying both requests and limits. It gets request=256Mi and limit=512Mi. Working as expected.
- When I create a pod with limit=x and unspecified request, it gets created with limit=x and request=x. Working as expected.
- When I create a pod with request=x and unspecified limit, it gets created with request=x and limit=512Mi. This is okay when x is less than 512Mi. **But when x is greater than 512Mi, I get an error**:
  ```
  The Pod "blah" is invalid: spec.containers[0].resources.limits: Invalid value: "512Mi": must be greater than or equal to memory request
  ```

**What you expected to happen**:
The default limit should not act as the max limit. It should have allowed me to create a pod with request greater than the default limit. I am not sure what limit value LimitRanger should set in this case, we can either set it equal to request or we can keep it unspecified.

However, if this is expected behavior, I should get a better error message that clearly says that if I am only specifying request, I need to specify it less than the default limit.

**Environment**:
- Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.4", GitCommit:"793658f2d7ca7f064d2bdf606519f9fe1229c381", GitTreeState:"clean", BuildDate:"2017-08-17T17:03:51Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.0", GitCommit:"d3ada0119e776222f11ec7945e6d860061339aad", GitTreeState:"clean", BuildDate:"2017-07-26T00:12:31Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
```
minikube version: v0.21.0
```

/sig scheduling
/cc @vishh

### PR DESCRIPTION
**Pull request by weiwei04**
**What this PR does / why we need it**:

the LimitRanger AdmissionController enforce two limit policies:
* **default resource policy**: set the default resource request/limit if pod.Spec.resource is empty

* **min/max resource policy**: make sure resource.request/limit is in limit range [min, max] range.

fixes: #51430 when enforce default resource policy, check `default.limit > request.resource`,  and return a better error message if the policy are violated.

**Which issue this PR fixes**: fixes #51430

**Special notes for your reviewer**:

**Release note**:

```release-note
none
```

### CODE REVIEW

**Comment by weiwei04** [comment]
@derekwaynecarr please take a look :)

File: plugin/pkg/admission/limitranger/admission.go [hunk 1]
```diff
@@ -212,6 +213,12 @@ func mergeContainerResources(container *api.Container, defaultRequirements *api.
 	for k, v := range defaultRequirements.Limits {
 		_, found := container.Resources.Limits[k]
 		if !found {
+			request, found := container.Resources.Requests[k]
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/limitranger/admission.go]: if the request is greater than the default limit, we should just not set the limit.

i do not see why we need to track errors.

File: plugin/pkg/admission/limitranger/admission.go [hunk 2]
```diff
@@ -233,20 +240,24 @@ func mergeContainerResources(container *api.Container, defaultRequirements *api.
...
 	}
-	return annotations
+	return annotations, errs
 }
 
 // mergePodResourceRequirements merges enumerated requirements with default requirements
 // it annotates the pod with information about what requirements were modified
-func mergePodResourceRequirements(pod *api.Pod, defaultRequirements *api.ResourceRequirements) {
+func mergePodResourceRequirements(pod *api.Pod, defaultRequirements *api.ResourceRequirements) []error {
+	subErrs := []error{}
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/limitranger/admission.go]: we dont need to track subErrs... just dont do the merge per prior comment.

File: plugin/pkg/admission/limitranger/admission_test.go [hunk 3]
```diff
@@ -250,6 +259,14 @@ func TestMergePodResourceRequirements(t *testing.T) {
 		}
 	}
 	expectNoAnnotation(t, &pod)
+
+	// pod with request enumerated and request.cpu > default.limit.cpu
+	input = getResourceRequirements(getComputeResourceList("100m", "512Mi"), getComputeResourceList("", ""))
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/limitranger/admission_test.go]: we shouldnt get an error per earlier comments.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
some updates requested throughout.

thanks for picking up the work here to make things better for users.

**Pull request closed without merging**